Sources/UrlFormEncoding/UrlFormDecoder.swift

```diff
@@ -0,0 +1,629 @@
+    }
+
+    func decode(_ type: UInt8.Type) throws -> UInt8 {
+      return try self.unwrap(UInt8.init)
+    }
+
+    func decode(_ type: UInt16.Type) throws -> UInt16 {
+      return try self.unwrap(UInt16.init)
+    }
+
+    func decode(_ type: UInt32.Type) throws -> UInt32 {
+      return try self.unwrap(UInt32.init)
+    }
+
+    func decode(_ type: UInt64.Type) throws -> UInt64 {
+      return try self.unwrap(UInt64.init)
+    }
+
+    func decode(_ type: Float.Type) throws -> Float {
+      return try self.unwrap(Float.init)
+    }
+
+    func decode(_ type: Double.Type) throws -> Double {
+      return try self.unwrap(Double.init)
+    }
+
+    func decode(_ type: String.Type) throws -> String {
+      return try self.unwrap(id)
+    }
+
+    func decode<T>(_ type: T.Type) throws -> T where T: Decodable {
+      self.decoder.containers.append(self.container)
+      defer { self.decoder.containers.removeLast() }
+      return try self.decoder.unbox(self.container, as: T.self)
+    }
+  }
+
+  public enum DataDecodingStrategy {
+    case deferredToData
+    case base64
+    case custom((String) -> Data?)
+  }
+
+  public enum DateDecodingStrategy {
+    case deferredToDate
+    case secondsSince1970
+    case millisecondsSince1970
+    case iso8601
+    case formatted(DateFormatter)
+    case custom((String) -> Date?)
+  }
+
+  public struct ParsingStrategy {
+    let strategy: (String) -> [String: Any]
+
+    public static let accumulatePairs = ParsingStrategy { query in
+      var params: [String: Any] = [:]
+      for (name, value) in pairs(query) {
+        var values = params[name] as? [Any] ?? []
+        values.append(value)
+        params[name] = values
+      }
+      return params
+    }
+
+    public static let brackets = ParsingStrategy.custom(parse(isArray: ^\.isEmpty))
+
+    public static let bracketsWithIndices = ParsingStrategy.custom(parse(isArray: { Int($0) != nil }))
+
+    public static let custom = ParsingStrategy.init
+  }
+}
+
+extension UrlFormDecoder.UnkeyedContainer.Key: CodingKey {
+  public var stringValue: String {
+    return String(self.index)
+  }
+
+  public init?(stringValue: String) {
+    guard let intValue = Int(stringValue) else { return nil }
+    self.init(intValue: intValue)
+  }
+
+  public var intValue: Int? {
+    return .some(self.index)
+  }
+
+  public init?(intValue: Int) {
+    self.init(index: intValue)
+  }
+}
+
+private let iso8601: (DateFormatter) -> DateFormatter = {
+  $0.calendar = Calendar(identifier: .iso8601)
+  $0.locale = Locale(identifier: "en_US_POSIX")
+  $0.timeZone = TimeZone(abbreviation: "UTC")
+  return $0
+}
+
+private let iso8601DateFormatter = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
+
+private let iso8601DateFormatterWithoutMilliseconds = DateFormatter()
+  |> iso8601
+  |> \.dateFormat .~ "yyyy-MM-dd'T'HH:mm:ssXXXXX"
+
+private func parse(isArray: @escaping (String) -> Bool) -> (String) -> [String: Any] {
```

Comment:
From here to the end of this file is an implementation of nested query parameter parsing. It's gnarly and a little imperative (I hope there aren't too many bugs!). Ideas for cleanup welcome!